include/seahorn/Expr/HexDump.hh

```diff
@@ -0,0 +1,78 @@
+
+#pragma once
+#include "seahorn/Expr/ExprOpCore.hh"
+#include "llvm/Support/raw_ostream.h"
+
+namespace expr {
+
+namespace hexDump {
+/**
+ * A Pair is an individual cell in the hex dump. It stores the index and value
+ * expressions
+ *
+ * \note bvnums and Uints are stored as MPZ expressions
+ */
+class Pair {
+  std::pair<Expr, Expr> m_pair; // <index, value>
+  bool m_isRepeated = false;
+
+public:
+  Pair(Expr idx, Expr value, bool isRepeated = false);
+
+  bool isSortable() const;
+
+  void print(std::ostream &OS, std::pair<unsigned, unsigned> widths,
+             bool includeAscii) const;
+
+  mpz_class getIdxNum() const;
+
+  Expr getIdxExpr() const;
+
+  Expr getValueExpr() const;
+
+  bool isRepeated() const;
+
+  /// based off of the index
+  bool operator<(Pair const &pair) const;
+
+  bool operator==(Pair const &pair) const;
+
+  /// prints with ascii and no set width
+  friend std::ostream &operator<<(std::ostream &OS, Pair const &pair);
+};
+
+using IteratorPair = std::pair<std::vector<Pair>::const_iterator,
+                               std::vector<Pair>::const_iterator>;
+
+/**
+ * For MK_STRUCT expressions, splits every child into a separate hex dump
+ *
+ * Supports nested ITE expressions, arrays (store expressions) and finite maps
+ * (set expressions)
+ *
+ * If the given expression is not one of the supported expressions above, it
+ * will search its children for the first supported expression and use that.
+ */
+class HexDump {
+  class Impl;
+  Impl *m_impl;
+
+public:
+  HexDump(Expr exp, unsigned addressesPerWord = 1);
+  ~HexDump();
+
+  /// return iterators to the first hex dump
+  std::vector<Pair>::const_iterator beginFirst() const;
+  std::vector<Pair>::const_iterator endFirst() const;
+
+  /// \return pairs of iterators to every hex dump
+  /// the pairs are of the following format: <begin, end>
+  std::vector<IteratorPair> getIterators() const;
```

Comment:
The goal that I had in mind with this was to give you all of the iterators for all the hex dumps (If you make a struct, I made it so that every child is a completely separate hex dump). This just gives you an std::pair with the begin and end iterators. Would it be better to just this a range instead?

Reply:
definitely. If you are returning a range use a range. A while back, pair of iterators was considered a range, but that is not true anymore. llvm has support for ranges.

Reply:
sounds good